src/service/viewer-impl.js

```diff
@@ -499,6 +521,55 @@ export class Viewer {
     return this.isWebviewEmbedded_;
   }
 
+  /**
+   * Whether the document is embedded in a Chrome Custom Tab.
+   * @return {boolean}
+   */
+  isCctEmbedded() {
+    return this.isCctEmbedded_;
+  }
+
+  /**
+   * Update the URL fragment with data needed to support custom tabs. This will
+   * not clear query string parameters, but will clear the fragment.
+   */
+  maybeUpdateFragmentForCct() {
+    if (!this.isCctEmbedded_) {
+      return;
+    }
+    // CCT only works with versions of Chrome that support the history API.
+    if (!this.win.history.replaceState) {
+      return;
+    }
+    const sourceUrl = getSourceUrl(this.win.location.href);
+    const canonicalUrl = Services.documentInfoForDoc(this.ampdoc).canonicalUrl;
+    const canonicalSourceUrl = getSourceUrl(canonicalUrl);
+    // Ensure that both origins are related.
+    if (this.getRootDomain_(sourceUrl) == this.getRootDomain_(canonicalSourceUrl)) {
+      const oldFragment = getFragment(this.win.location.href);
+      const newFragment = 'ampshare=' + encodeURIComponent(canonicalUrl);
+      // Attempt to merge the fragments, if an old fragment was present.
+      this.win.history.replaceState({}, '',
+          oldFragment ? `${oldFragment}&${newFragment}` : `#${newFragment}`);
+    }
+  }
+
+  /**
+   * Removes extraneous subdomains from a URL for comparison.
+   * @param {string} url
+   * @return {string} The root domain.
+   * @private
+   */
+  getRootDomain_(url) {
+    let hostname = parseUrl(url).hostname;
+    // TODO: this should compute the root domain, and not just remove common
+    // subdomains.
```

Comment:
Do we plan to make this change? If not, let's just remove the TODO.

Then we can rename this function to be more precise, e.g. 

```js
/** @return {boolean} */
isRoughlySameOrigin_(a, b) {
}
```

Reply:
Good point -- it does seem like adding this will be quite involved and may not even happen. I'll update!